File: visual-studio-extension/DeepLensVisualStudio/DeepLensVisualStudio/Services/SlashCommandService.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace DeepLensVisualStudio.Services
{
    public enum SlashCommandCategory
    {
        Search,
        Navigation,
        Files,
        Refactoring,
        Actions
    }

    public class SlashCommand
    {
        public string Name { get; set; } = "";
        public string ShortName { get; set; } = "";
        public string Description { get; set; } = "";
        public string[] Aliases { get; set; } = Array.Empty<string>();
        public SlashCommandCategory Category { get; set; }
        public string Icon { get; set; } = "";
        public string KeyboardShortcut { get; set; } = "";
        public string Example { get; set; } = "";
    }

    public class SlashCommandService
    {
        private readonly List<SlashCommand> _commands;
        private readonly Dictionary<SlashCommandCategory, List<SlashCommand>> _categoryGroups;
        private readonly List<string> _recentlyUsed;

        public SlashCommandService()
        {
            _commands = new List<SlashCommand>
            {
                new SlashCommand 
                { 
                    Name = "/all", 
                    ShortName = "all",
                    Description = "Search everywhere across all scopes",
                    Aliases = new[] { "/a" },
                    Category = SlashCommandCategory.Search,
                    Icon = "search",
                    KeyboardShortcut = "Ctrl+T",
                    Example = "/all UserService"
                },
                new SlashCommand 
                { 
                    Name = "/t", 
                    ShortName = "t",
                    Description = "Find classes, interfaces, and types",
                    Aliases = new[] { "/classes", "/types", "/type", "/c" },
                    Category = SlashCommandCategory.Search,
                    Icon = "symbol-class",
                    KeyboardShortcut = "Ctrl+Shift+T",
                    Example = "/t User"
                },
                new SlashCommand 
                { 
                    Name = "/s", 
                    ShortName = "s",
                    Description = "Find methods, functions, and variables",
                    Aliases = new[] { "/symbols", "/symbol", "#" },
                    Category = SlashCommandCategory.Search,
                    Icon = "symbol-method",
                    KeyboardShortcut = "Ctrl+Shift+O",
                    Example = "/s getUsers"
                },
                new SlashCommand 
                { 
                    Name = "/f", 
                    ShortName = "f",
                    Description = "Find files by name or path",
                    Aliases = new[] { "/files", "/file" },
                    Category = SlashCommandCategory.Files,
                    Icon = "file",
                    KeyboardShortcut = "Ctrl+Shift+F",
                    Example = "/f UserService.cs"
                },
                new SlashCommand 
                { 
                    Name = "/txt", 
                    ShortName = "txt",
                    Description = "Search text content across files",
                    Aliases = new[] { "/text", "/find", "/grep" },
                    Category = SlashCommandCategory.Search,
                    Icon = "whole-word",
                    KeyboardShortcut = "Ctrl+Shift+G",
                    Example = "/txt \"async function\""
                },
                new SlashCommand 
                { 
                    Name = "/e", 
                    ShortName = "e",
                    Description = "Find API endpoints and routes",
                    Aliases = new[] { "/endpoints", "/endpoint", "/routes", "/api" },
                    Category = SlashCommandCategory.Search,
                    Icon = "globe",
                    Example = "/e /api/users"
                },
                new SlashCommand 
                { 
                    Name = "/o", 
                    ShortName = "o",
                    Description = "Search in currently open files",
                    Aliases = new[] { "/open", "/opened" },
                    Category = SlashCommandCategory.Navigation,
                    Icon = "book",
                    Example = "/o search"
                },
                new SlashCommand 
                { 
                    Name = "/m", 
                    ShortName = "m",
                    Description = "Search in modified or untracked files",
                    Aliases = new[] { "/modified", "/mod", "/git", "/changed" },
                    Category = SlashCommandCategory.Files,
                    Icon = "git-merge",
                    Example = "/m service"
                },
                new SlashCommand 
                { 
                    Name = "/p", 
                    ShortName = "p",
                    Description = "Find properties and fields",
                    Aliases = new[] { "/properties", "/prop", "/field" },
                    Category = SlashCommandCategory.Search,
                    Icon = "symbol-property",
                    Example = "/p userId"
                },
                new SlashCommand 
                { 
                    Name = "/cmd", 
                    ShortName = "cmd",
                    Description = "Search and execute VS commands",
                    Aliases = new[] { "/commands", "/action", "/run", ">" },
                    Category = SlashCommandCategory.Actions,
                    Icon = "run",
                    KeyboardShortcut = "Ctrl+Shift+P",
                    Example = "/cmd \"format document\""
                },
            };

            _categoryGroups = _commands
                .GroupBy(c => c.Category)
                .ToDictionary(g => g.Key, g => g.ToList());

            _recentlyUsed = LoadRecentCommands();
        }

        public IEnumerable<SlashCommand> GetCommands(string query = "")
        {
            if (string.IsNullOrEmpty(query))
                return _commands;

            string lowerQuery = query.ToLowerInvariant();
            var results = new List<SlashCommand>();
            var seen = new HashSet<string>();

            foreach (var cmd in _commands)
            {
                if (seen.Contains(cmd.Name)) continue;

                bool exactMatch = cmd.Name.Equals(lowerQuery, StringComparison.OrdinalIgnoreCase);
                bool startsWithMatch = cmd.Name.StartsWith(lowerQuery, StringComparison.OrdinalIgnoreCase);
                bool aliasMatch = cmd.Aliases.Any(a => 
                    a.Equals(lowerQuery, StringComparison.OrdinalIgnoreCase) || 
                    a.StartsWith(lowerQuery, StringComparison.OrdinalIgnoreCase));
                bool descriptionMatch = cmd.Description.IndexOf(lowerQuery, StringComparison.OrdinalIgnoreCase) >= 0;
                bool shortNameMatch = cmd.ShortName.StartsWith(lowerQuery, StringComparison.OrdinalIgnoreCase);

                if (exactMatch || startsWithMatch || aliasMatch || descriptionMatch || shortNameMatch)
                {
                    results.Add(cmd);
                    seen.Add(cmd.Name);
                }
            }

            SortResults(results, lowerQuery);
            return results;
        }

        public IEnumerable<SlashCommand> GetRecentCommands()
        {
            var recent = new List<SlashCommand>();
            var seen = new HashSet<string>();

            foreach (var name in _recentlyUsed)
            {
                var cmd = _commands.FirstOrDefault(c => 
                    c.Name.Equals(name, StringComparison.OrdinalIgnoreCase) ||
                    c.Aliases.Any(a => a.Equals(name, StringComparison.OrdinalIgnoreCase)));
                
                if (cmd != null && !seen.Contains(cmd.Name))
                {
                    recent.Add(cmd);
                    seen.Add(cmd.Name);
                }
            }

            return recent;
        }

        public IEnumerable<SlashCommand> GetCommandsByCategory(SlashCommandCategory category)
        {
            return _categoryGroups.ContainsKey(category) 
                ? _categoryGroups[category] 
                : Enumerable.Empty<SlashCommand>();
        }

        public IEnumerable<SlashCommandCategory> GetAllCategories()
        {
            return _categoryGroups.Keys;
        }

        public SlashCommand GetCommand(string commandOrAlias)
        {
            string normalized = commandOrAlias.ToLowerInvariant();
            return _commands.FirstOrDefault(c => 
                c.Name.Equals(normalized, StringComparison.OrdinalIgnoreCase) ||
                c.Aliases.Any(a => a.Equals(normalized, StringComparison.OrdinalIgnoreCase)));
        }

        public string GetPrimaryAlias(SlashCommand cmd)
        {
            return cmd.Aliases.Length > 0 ? cmd.Aliases[0] : cmd.Name;
        }

        public void RecordUsage(string commandName)
        {
            string normalized = commandName.ToLowerInvariant();
            _recentlyUsed.RemoveAll(c => c.Equals(normalized, StringComparison.OrdinalIgnoreCase));
            _recentlyUsed.Insert(0, normalized);
            
            if (_recentlyUsed.Count > 10)
                _recentlyUsed.RemoveAt(_recentlyUsed.Count - 1);

            SaveRecentCommands();
        }

        private void SortResults(List<SlashCommand> results, string query)
        {
            string lowerQuery = query.ToLowerInvariant();

            results.Sort((a, b) =>
            {
                bool aExact = a.Name.Equals(lowerQuery, StringComparison.OrdinalIgnoreCase);
                bool bExact = b.Name.Equals(lowerQuery, StringComparison.OrdinalIgnoreCase);

                if (aExact && !bExact) return -1;
                if (!aExact && bExact) return 1;

                bool aRecent = _recentlyUsed.Any(r => 
                    r.Equals(a.Name, StringComparison.OrdinalIgnoreCase) ||
                    a.Aliases.Any(ali => ali.Equals(r, StringComparison.OrdinalIgnoreCase)));
                bool bRecent = _recentlyUsed.Any(r => 
                    r.Equals(b.Name, StringComparison.OrdinalIgnoreCase) ||
                    b.Aliases.Any(ali => ali.Equals(r, StringComparison.OrdinalIgnoreCase)));

                if (aRecent && !bRecent) return -1;
                if (!aRecent && bRecent) return 1;

                bool aStartsWith = a.Name.StartsWith(lowerQuery, StringComparison.OrdinalIgnoreCase);
                bool bStartsWith = b.Name.StartsWith(lowerQuery, StringComparison.OrdinalIgnoreCase);

                if (aStartsWith && !bStartsWith) return -1;
                if (!aStartsWith && bStartsWith) return 1;

                return string.Compare(a.Name, b.Name, StringComparison.Ordinal);
            });
        }

        private List<string> LoadRecentCommands()
        {
            try
            {
                var settings = Microsoft.VisualStudio.Shell.Package.GetGlobalService(typeof(Microsoft.VisualStudio.Shell.Interop.SVsSettingsManager)) 
                    as Microsoft.VisualStudio.Shell.Interop.IVsSettingsManager;
                
                // This would need proper settings storage implementation
                // For now, return empty list
                return new List<string>();
            }
            catch
            {
                return new List<string>();
            }
        }

        private void SaveRecentCommands()
        {
            try
            {
                // This would need proper settings storage implementation
            }
            catch
            {
                // Ignore save errors
            }
        }

        public string FormatCommandForDisplay(SlashCommand cmd)
        {
            string shortcuts = !string.IsNullOrEmpty(cmd.KeyboardShortcut) ? $" [{cmd.KeyboardShortcut}]" : "";
            return $"{GetPrimaryAlias(cmd)}: {cmd.Description}{shortcuts}";
        }

        public string GetCategoryLabel(SlashCommandCategory category)
        {
            switch (category)
            {
                case SlashCommandCategory.Search: return "Search";
                case SlashCommandCategory.Navigation: return "Navigation";
                case SlashCommandCategory.Files: return "Files";
                case SlashCommandCategory.Refactoring: return "Refactoring";
                case SlashCommandCategory.Actions: return "Actions";
                default: return category.ToString();
            }
        }
    }
}

File: vscode-extension/src/slash-command-service.ts
import * as vscode from 'vscode';
import { SearchScope } from '../../language-server/src/core/types';

export interface SlashCommand {
    id: string;
    name: string;
    shortName: string;
    description: string;
    aliases: string[];
    category: SlashCommandCategory;
    scope: SearchScope;
    icon: string;
    keyboardShortcut?: string;
    example?: string;
}

export enum SlashCommandCategory {
    SEARCH = 'Search',
    NAVIGATION = 'Navigation',
    FILES = 'Files',
    REFACTORING = 'Refactoring',
    ACTIONS = 'Actions',
}

export class SlashCommandService {
    private readonly commands: Map<string, SlashCommand> = new Map();
    private readonly categoryGroups: Map<SlashCommandCategory, SlashCommand[]> = new Map();
    private recentlyUsed: string[] = [];

    constructor() {
        this.initializeCommands();
        this.categoryGroups = this.initializeGroups();
        this.loadRecentCommands();
    }

    private initializeCommands(): void {
        const commands: SlashCommand[] = [
            {
                id: '/all',
                name: '/all',
                shortName: 'all',
                description: 'Search everywhere across all scopes',
                aliases: ['/a'],
                category: SlashCommandCategory.SEARCH,
                scope: SearchScope.EVERYTHING,
                icon: 'search',
                keyboardShortcut: 'Ctrl+T',
                example: '/all UserService',
            },
            {
                id: '/t',
                name: '/t',
                shortName: 't',
                description: 'Find classes, interfaces, and types',
                aliases: ['/classes', '/types', '/type', '/c'],
                category: SlashCommandCategory.SEARCH,
                scope: SearchScope.TYPES,
                icon: 'symbol-class',
                keyboardShortcut: 'Ctrl+Shift+T',
                example: '/t User',
            },
            {
                id: '/s',
                name: '/s',
                shortName: 's',
                description: 'Find methods, functions, and variables',
                aliases: ['/symbols', '/symbol', '#'],
                category: SlashCommandCategory.SEARCH,
                scope: SearchScope.SYMBOLS,
                icon: 'symbol-method',
                keyboardShortcut: 'Ctrl+Shift+O',
                example: '/s getUsers',
            },
            {
                id: '/f',
                name: '/f',
                shortName: 'f',
                description: 'Find files by name or path',
                aliases: ['/files', '/file'],
                category: SlashCommandCategory.FILES,
                scope: SearchScope.FILES,
                icon: 'file',
                keyboardShortcut: 'Ctrl+Shift+F',
                example: '/f UserService.ts',
            },
            {
                id: '/txt',
                name: '/txt',
                shortName: 'txt',
                description: 'Search text content across files',
                aliases: ['/text', '/find', '/grep'],
                category: SlashCommandCategory.SEARCH,
                scope: SearchScope.TEXT,
                icon: 'whole-word',
                keyboardShortcut: 'Ctrl+Shift+G',
                example: '/txt "async function"',
            },
            {
                id: '/e',
                name: '/e',
                shortName: 'e',
                description: 'Find API endpoints and routes',
                aliases: ['/endpoints', '/endpoint', '/routes', '/api'],
                category: SlashCommandCategory.SEARCH,
                scope: SearchScope.ENDPOINTS,
                icon: 'globe',
                example: '/e /api/users',
            },
            {
                id: '/o',
                name: '/o',
                shortName: 'o',
                description: 'Search in currently open files',
                aliases: ['/open', '/opened'],
                category: SlashCommandCategory.NAVIGATION,
                scope: SearchScope.OPEN,
                icon: 'book',
                example: '/o search',
            },
            {
                id: '/m',
                name: '/m',
                shortName: 'm',
                description: 'Search in modified or untracked files',
                aliases: ['/modified', '/mod', '/git', '/changed'],
                category: SlashCommandCategory.FILES,
                scope: SearchScope.MODIFIED,
                icon: 'git-merge',
                example: '/m service',
            },
            {
                id: '/p',
                name: '/p',
                shortName: 'p',
                description: 'Find properties and fields',
                aliases: ['/properties', '/prop', '/field'],
                category: SlashCommandCategory.SEARCH,
                scope: SearchScope.PROPERTIES,
                icon: 'symbol-property',
                example: '/p userId',
            },
            {
                id: '/cmd',
                name: '/cmd',
                shortName: 'cmd',
                description: 'Search and execute VS Code commands',
                aliases: ['/commands', '/action', '/run', '>'],
                category: SlashCommandCategory.ACTIONS,
                scope: SearchScope.COMMANDS,
                icon: 'run',
                keyboardShortcut: 'Ctrl+Shift+P',
                example: '/cmd "format document"',
            },
        ];

        for (const cmd of commands) {
            this.commands.set(cmd.name.toLowerCase(), cmd);
            this.commands.set(cmd.id, cmd);
            for (const alias of cmd.aliases) {
                this.commands.set(alias.toLowerCase(), cmd);
            }
        }
    }

    private initializeGroups(): Map<SlashCommandCategory, SlashCommand[]> {
        const grouped = new Map<SlashCommandCategory, SlashCommand[]>();
        const uniqueCommands = new Set<string>();

        for (const cmd of this.commands.values()) {
            if (!uniqueCommands.has(cmd.name)) {
                let group = grouped.get(cmd.category);
                if (!group) {
                    group = [];
                    grouped.set(cmd.category, group);
                }
                group.push(cmd);
                uniqueCommands.add(cmd.name);
            }
        }

        return grouped;
    }

    private loadRecentCommands(): void {
        const context = vscode.workspace.getConfiguration('deeplens');
        const recent = context.get<string[]>('recentSlashCommands', []);
        this.recentlyUsed = recent;
    }

    private async saveRecentCommands(): Promise<void> {
        const context = vscode.workspace.getConfiguration('deeplens');
        await context.update('recentSlashCommands', this.recentlyUsed, vscode.ConfigurationTarget.Global);
    }

    private recordUsage(commandName: string): void {
        const normalized = commandName.toLowerCase();
        this.recentlyUsed = this.recentlyUsed.filter((cmd) => cmd !== normalized);
        this.recentlyUsed.unshift(normalized);
        this.recentlyUsed = this.recentlyUsed.slice(0, 10);
        this.saveRecentCommands();
    }

    getCommands(query?: string): SlashCommand[] {
        if (!query) {
            return Array.from(new Set(Array.from(this.commands.values()).map((c) => c.name)))
                .map((name) => this.commands.get(name))
                .filter((cmd): cmd is SlashCommand => cmd !== undefined);
        }

        const lowerQuery = query.toLowerCase();
        const trimmedQuery = lowerQuery.replace(/^[/#>]/, '');
        const results: SlashCommand[] = [];
        const seen = new Set<string>();

        for (const cmd of this.commands.values()) {
            if (seen.has(cmd.name)) continue;

            const exactMatch = cmd.name === lowerQuery;
            const startsWithMatch = cmd.name.startsWith(lowerQuery);
            const aliasMatch = cmd.aliases.some(
                (alias) => alias.toLowerCase() === lowerQuery || alias.toLowerCase().startsWith(lowerQuery),
            );
            const descriptionMatch = cmd.description.toLowerCase().includes(trimmedQuery);
            const shortNameMatch = cmd.shortName.toLowerCase().startsWith(trimmedQuery);

            if (exactMatch || startsWithMatch || aliasMatch || descriptionMatch || shortNameMatch) {
                results.push(cmd);
                seen.add(cmd.name);
            }
        }

        this.sortResults(results, lowerQuery);
        return results;
    }

    getRecentCommands(): SlashCommand[] {
        const recent: SlashCommand[] = [];
        const seen = new Set<string>();

        for (const name of this.recentlyUsed) {
            const cmd = this.commands.get(name);
            if (cmd && !seen.has(cmd.name)) {
                recent.push(cmd);
                seen.add(cmd.name);
            }
        }

        return recent;
    }

    getCommandsByCategory(category: SlashCommandCategory): SlashCommand[] {
        return this.categoryGroups.get(category) || [];
    }

    getAllCategories(): SlashCommandCategory[] {
        return Array.from(this.categoryGroups.keys());
    }

    getCommand(commandOrAlias: string): SlashCommand | undefined {
        const normalized = commandOrAlias.toLowerCase();
        return this.commands.get(normalized);
    }

    executeCommand(commandName: string): void {
        this.recordUsage(commandName);
    }

    private sortResults(results: SlashCommand[], query: string): void {
        const lowerQuery = query.toLowerCase();

        results.sort((a, b) => {
            const aExact = a.name === lowerQuery;
            const bExact = b.name === lowerQuery;

            if (aExact && !bExact) return -1;
            if (!aExact && bExact) return 1;

            const aRecent = this.recentlyUsed.includes(a.name.toLowerCase());
            const bRecent = this.recentlyUsed.includes(b.name.toLowerCase());

            if (aRecent && !bRecent) return -1;
            if (!aRecent && bRecent) return 1;

            const aStartsWith = a.name.startsWith(lowerQuery);
            const bStartsWith = b.name.startsWith(lowerQuery);

            if (aStartsWith && !bStartsWith) return -1;
            if (!aStartsWith && bStartsWith) return 1;

            return a.name.localeCompare(b.name);
        });
    }

    formatCommandForDisplay(cmd: SlashCommand): string {
        const primaryAlias = cmd.aliases[0] || cmd.name;
        const shortcuts = cmd.keyboardShortcut ? ` [${cmd.keyboardShortcut}]` : '';
        return `${primaryAlias}: ${cmd.description}${shortcuts}`;
    }

    getPrimaryAlias(cmd: SlashCommand): string {
        return cmd.name;
    }

    getAliasesForDisplay(cmd: SlashCommand): string[] {
        const primaryAlias = this.getPrimaryAlias(cmd);
        const aliases = [primaryAlias, cmd.name, ...cmd.aliases];
        const seen = new Set<string>();

        return aliases.filter((alias) => {
            const normalized = alias.toLowerCase();
            if (seen.has(normalized)) {
                return false;
            }
            seen.add(normalized);
            return true;
        });
    }

    getCategoryIcon(category: SlashCommandCategory): string {
        switch (category) {
            case SlashCommandCategory.SEARCH:
                return 'search';
            case SlashCommandCategory.NAVIGATION:
                return 'arrow-right';
            case SlashCommandCategory.FILES:
                return 'file';
            case SlashCommandCategory.REFACTORING:
                return 'wand';
            case SlashCommandCategory.ACTIONS:
                return 'run';
            default:
                return 'circle-outline';
        }
    }
}
